src/Http/WebUtilities/src/MultipartPipeReader.cs

```diff
@@ -0,0 +1,330 @@
+                ref accumulator,
+                ref headersLength,
+                isFinalBlock);
+        }
+
+        // Fast parsing for single span in ReadOnlySequence
+        private bool TryParseHeadersToEndFast(ReadOnlySpan<byte> span,
+            ref KeyValueAccumulator accumulator,
+            bool isFinalBlock,
+            long lengthLimit,
+            out int consumed)
+        {
+            consumed = 0;
+
+            while (span.Length > 0)
+            {
+                // Find the end of the header.
+                var newLine = span.IndexOf(CrlfDelimiter);
+                ReadOnlySpan<byte> line;
+                var foundNewLine = newLine != -1;
+
+                if (foundNewLine)
+                {
+                    line = span.Slice(0, newLine);
+                    span = span.Slice(line.Length + CrlfDelimiter.Length);
+                    consumed += line.Length + CrlfDelimiter.Length;
+                }
+                // We can't know that what is currently read is the end of the header value, that's only the case if this is the final block
+                // If we're not in the final block, then consume nothing
+                else if (!isFinalBlock)
+                {
+                    // Don't buffer indefinately
+                    if (span.Length > lengthLimit)
+                    {
+                        throw new InvalidDataException($"Line length limit {lengthLimit} exceeded.");
+                    }
+                    return false;
+                }
+                else
+                {
+                    throw new InvalidDataException("Unexpected end of Stream, the content may have already been read by another component. ");
+                }
+
+                if (line.Length == 0) // an empty line means it's the end of the headers
+                {
+                    return true;
+                }
+
+                if (line.Length > lengthLimit)
+                {
+                    throw new InvalidDataException($"Line length limit {lengthLimit} exceeded.");
+                }
+
+                ParseHeaderLineFast(line, ref accumulator);
+                lengthLimit -= line.Length;
+            }
+            return false;
+        }
+
+
+        private void ParseHeaderLineFast(ReadOnlySpan<byte> line,
+          ref KeyValueAccumulator accumulator)
+        {
+            ReadOnlySpan<byte> key;
+            ReadOnlySpan<byte> value;
+
+            int colon = line.IndexOf(ColonDelimiter);
+
+            if (colon == -1)
+            {
+                throw new InvalidDataException($"Invalid header line: {GetDecodedString(line)}");
+            }
+            else
+            {
+                key = line.Slice(0, colon);
+                value = line.Slice(colon + ColonDelimiter.Length);
+            }
+            var decodedKey = GetDecodedString(key).Trim();
+            var decodedValue = GetDecodedString(value).Trim();
```

Comment:
Perf: Should find a way to incorporate the trim logic into GetDecodedString to avoid the double allocation. The value will almost always have a space at the beginning after the colon separator. Not a blocker on this PR, but worth looking at.